plutus-metatheory/test/TestNEAT.hs

```diff
@@ -1,105 +1,260 @@
+  tyN1 <- withExceptT Ctrex $
+    case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
+      Just tyN -> return tyN
+      Nothing  -> throwError (CtrexTypeNormalizationFail k tyG)
+  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
+  ty2 <- withExceptT TypeError $ unNormalized <$> normalizeType ty
+  unless ((() <$ ty1) == ty2) $
+    throwCtrex (CtrexTypeNormalizationMismatch k tyG (() <$ ty1) ty2)
 
-star :: Kind ()
-star = Type ()
+-- compare the production kind inference against the Agda
+prop_kindInferSame :: Kind ()
+                   -> ClosedTypeG
+                   -> ExceptT TestFail Quote ()
+prop_kindInferSame k tyG = do
+  ty <- withExceptT GenError $ convertClosedType tynames k tyG
+  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
+  k' <- withExceptT Ctrex $ case inferKindAgda (AlexPn 0 0 0 <$ tyDB) of
+    Just k' -> return k'
+    Nothing -> throwError (CtrexKindCheckFail k tyG)
+  k'' <- withExceptT TypeError $ inferKind defConfig (() <$ ty)
+  unless (unconvK (unDeBruijnifyK k') == k'') $ throwCtrex (CtrexKindMismatch k tyG (unconvK (unDeBruijnifyK k')) k'')
 
--- |Check if the type/kind checker or generation threw any errors.
-isSafe :: ExceptT (ErrorP a) Quote a -> Cool
-isSafe = toCool . isRight . runQuote . runExceptT
+-- try to infer the type of a term
+prop_typeinfer :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_typeinfer (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  withExceptT Ctrex $
+    case inferTypeAgda (AlexPn 0 0 0 <$ tmDB) of
+      Just _  -> return ()
+      Nothing -> throwError (CtrexTypeCheckFail tyG tmG)
 
-prop_checkKindSound :: TyProp
-prop_checkKindSound k _ tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
+-- try to typecheck a term
+prop_typecheck :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_typecheck (k , tyG) tmG = do
+  ty <- withExceptT GenError $ convertClosedType tynames k tyG
   tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  withExceptT AgdaErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyDB) (deBruijnifyK (convK k)) of
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  case checkTypeAgda (AlexPn 0 0 0 <$ tyDB) (AlexPn 0 0 0 <$ tmDB) of
     Just _  -> return ()
-    Nothing -> throwError ()
+    Nothing -> throwCtrex (CtrexTypeCheckFail tyG tmG)
 
-prop_normalizePreservesKind :: TyProp
-prop_normalizePreservesKind k _ tyQ = isSafe $ do
-  ty  <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  withExceptT AgdaErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyN) (deBruijnifyK (convK k)) of
+prop_runCK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
```

Comment:
Error: It does not generate `error` yet. I think it should still work if it did. 

Looping: I guess it would need to generated the Z-combinator but it doesn't generate terms to that depth so it nothing loops.

I have merged this with other tests.

Reply:
> Looping: I guess it would need to generated the Z-combinator but it doesn't generate terms to that depth so it nothing loops.

How big is the Omega combinator?